File: snippets/security/authentication/bearer_token_auth.ts
/**
 * Bearer Token Authentication (TypeScript)
 * Production-ready Bearer token implementation for API authentication
 */

interface User {
  user_id: number;
  username: string;
  scopes: string[];
  [key: string]: any;
}

type TokenValidator = (token: string) => Promise<User | null> | User | null;

export class BearerTokenAuth {
  private tokenValidator: TokenValidator;

  /**
   * Initialize bearer token auth
   * @param tokenValidator - Function to validate token and return user data
   */
  constructor(tokenValidator: TokenValidator) {
    this.tokenValidator = tokenValidator;
  }

  /**
   * Extract bearer token from Authorization header
   * @param authorizationHeader - Authorization header value
   * @returns Token string or null
   */
  extractToken(authorizationHeader?: string): string | null {
    if (!authorizationHeader) {
      return null;
    }

    // Bearer token format: "Bearer <token>"
    const match = authorizationHeader.match(/^Bearer\s+(.+)$/i);

    if (match) {
      return match[1];
    }

    return null;
  }

  /**
   * Authenticate request using bearer token
   * @param authorizationHeader - Authorization header value
   * @returns User data if authenticated, null otherwise
   */
  async authenticate(authorizationHeader?: string): Promise<User | null> {
    const token = this.extractToken(authorizationHeader);

    if (!token) {
      return null;
    }

    // Validate token
    return await this.tokenValidator(token);
  }

  /**
   * Check if user has required scope
   * @param user - User object
   * @param requiredScope - Required permission scope
   * @returns True if user has scope
   */
  hasScope(user: User, requiredScope: string): boolean {
    return user.scopes.includes(requiredScope);
  }

  /**
   * Check if user has any of the required scopes
   * @param user - User object
   * @param requiredScopes - Array of required scopes
   * @returns True if user has at least one scope
   */
  hasAnyScope(user: User, requiredScopes: string[]): boolean {
    return requiredScopes.some((scope) => user.scopes.includes(scope));
  }

  /**
   * Check if user has all required scopes
   * @param user - User object
   * @param requiredScopes - Array of required scopes
   * @returns True if user has all scopes
   */
  hasAllScopes(user: User, requiredScopes: string[]): boolean {
    return requiredScopes.every((scope) => user.scopes.includes(scope));
  }
}

/**
 * Express middleware for Bearer token authentication
 */
export class ExpressBearerAuth {
  private auth: BearerTokenAuth;

  constructor(tokenValidator: TokenValidator) {
    this.auth = new BearerTokenAuth(tokenValidator);
  }

  /**
   * Middleware to require authentication
   */
  requireAuth() {
    return async (req: any, res: any, next: any) => {
      const authHeader = req.headers.authorization;
      const user = await this.auth.authenticate(authHeader);

      if (!user) {
        return res.status(401).json({
          error: 'Unauthorized',
          message: 'Valid authentication required',
        });
      }

      req.user = user;
      next();
    };
  }

  /**
   * Middleware to require specific scope
   * @param requiredScope - Required permission scope
   */
  requireScope(requiredScope: string) {
    return async (req: any, res: any, next: any) => {
      const authHeader = req.headers.authorization;
      const user = await this.auth.authenticate(authHeader);

      if (!user) {
        return res.status(401).json({
          error: 'Unauthorized',
          message: 'Valid authentication required',
        });
      }

      if (!this.auth.hasScope(user, requiredScope)) {
        return res.status(403).json({
          error: 'Forbidden',
          message: 'Insufficient permissions',
        });
      }

      req.user = user;
      next();
    };
  }

  /**
   * Middleware to require any of the specified scopes
   * @param requiredScopes - Array of required scopes
   */
  requireAnyScope(requiredScopes: string[]) {
    return async (req: any, res: any, next: any) => {
      const authHeader = req.headers.authorization;
      const user = await this.auth.authenticate(authHeader);

      if (!user) {
        return res.status(401).json({
          error: 'Unauthorized',
          message: 'Valid authentication required',
        });
      }

      if (!this.auth.hasAnyScope(user, requiredScopes)) {
        return res.status(403).json({
          error: 'Forbidden',
          message: 'Insufficient permissions',
        });
      }

      req.user = user;
      next();
    };
  }
}

/**
 * FastAPI-style dependency for Bearer auth
 */
export class FastAPIBearerAuth {
  private auth: BearerTokenAuth;

  constructor(tokenValidator: TokenValidator) {
    this.auth = new BearerTokenAuth(tokenValidator);
  }

  /**
   * Get current authenticated user
   * @param authHeader - Authorization header
   * @returns User or throws error
   */
  async getCurrentUser(authHeader?: string): Promise<User> {
    const user = await this.auth.authenticate(authHeader);

    if (!user) {
      throw new Error('Unauthorized');
    }

    return user;
  }

  /**
   * Get current user with required scope
   * @param authHeader - Authorization header
   * @param requiredScope - Required scope
   * @returns User or throws error
   */
  async getCurrentUserWithScope(
    authHeader: string | undefined,
    requiredScope: string
  ): Promise<User> {
    const user = await this.getCurrentUser(authHeader);

    if (!this.auth.hasScope(user, requiredScope)) {
      throw new Error('Forbidden');
    }

    return user;
  }
}

// Example usage
if (require.main === module) {
  // Create token validator
  const validateToken = async (token: string): Promise<User | null> => {
    // In production, validate JWT or query database
    const validTokens: Record<string, User> = {
      secret_token_123: {
        user_id: 1,
        username: 'john_doe',
        scopes: ['read', 'write', 'admin'],
      },
    };

    return validTokens[token] || null;
  };

  // Initialize auth
  const auth = new BearerTokenAuth(validateToken);

  // Test authentication
  (async () => {
    const authHeader = 'Bearer secret_token_123';
    const user = await auth.authenticate(authHeader);

    if (user) {
      console.log('Authenticated:', user);
      console.log('Has admin scope:', auth.hasScope(user, 'admin'));
      console.log(
        'Has any scope:',
        auth.hasAnyScope(user, ['admin', 'superuser'])
      );
    } else {
      console.log('Authentication failed');
    }

    // Test invalid token
    const invalidUser = await auth.authenticate('Bearer invalid_token');
    console.log('Invalid token result:', invalidUser);
  })();
}

File: snippets/security/authentication/bearer_token_auth.py
"""
Bearer Token Authentication
Production-ready Bearer token implementation for API authentication
"""
from typing import Optional, Callable, Dict
from functools import wraps
import re


class BearerTokenAuth:
    """Bearer token authentication handler"""

    def __init__(self, token_validator: Callable[[str], Optional[Dict]]):
        """
        Initialize bearer token auth

        Args:
            token_validator: Function to validate token and return user data
        """
        self.token_validator = token_validator

    def extract_token(self, authorization_header: Optional[str]) -> Optional[str]:
        """
        Extract bearer token from Authorization header

        Args:
            authorization_header: Authorization header value

        Returns:
            Token string or None
        """
        if not authorization_header:
            return None

        # Bearer token format: "Bearer <token>"
        match = re.match(r'^Bearer\s+(.+)$', authorization_header, re.IGNORECASE)

        if match:
            return match.group(1)

        return None

    def authenticate(self, authorization_header: Optional[str]) -> Optional[Dict]:
        """
        Authenticate request using bearer token

        Args:
            authorization_header: Authorization header value

        Returns:
            User data if authenticated, None otherwise
        """
        token = self.extract_token(authorization_header)

        if not token:
            return None

        # Validate token
        return self.token_validator(token)

    def require_auth(self, func):
        """
        Decorator to require authentication

        Usage:
            @auth.require_auth
            def protected_route(user):
                return f"Hello {user['username']}"
        """
        @wraps(func)
        def wrapper(request, *args, **kwargs):
            auth_header = request.headers.get('Authorization')
            user = self.authenticate(auth_header)

            if not user:
                return {'error': 'Unauthorized'}, 401

            return func(user, *args, **kwargs)

        return wrapper

    def require_scope(self, required_scope: str):
        """
        Decorator to require specific scope

        Args:
            required_scope: Required permission scope

        Usage:
            @auth.require_scope('admin')
            def admin_route(user):
                return "Admin panel"
        """
        def decorator(func):
            @wraps(func)
            def wrapper(request, *args, **kwargs):
                auth_header = request.headers.get('Authorization')
                user = self.authenticate(auth_header)

                if not user:
                    return {'error': 'Unauthorized'}, 401

                user_scopes = user.get('scopes', [])
                if required_scope not in user_scopes:
                    return {'error': 'Forbidden'}, 403

                return func(user, *args, **kwargs)

            return wrapper
        return decorator


class APITokenMiddleware:
    """Middleware for API token authentication"""

    def __init__(self, token_auth: BearerTokenAuth):
        """
        Initialize middleware

        Args:
            token_auth: BearerTokenAuth instance
        """
        self.token_auth = token_auth

    def process_request(self, request) -> Optional[Dict]:
        """
        Process incoming request for authentication

        Args:
            request: HTTP request object

        Returns:
            Error response or None to continue
        """
        # Skip authentication for public endpoints
        public_paths = ['/health', '/login', '/register']

        if request.path in public_paths:
            return None

        # Authenticate request
        auth_header = request.headers.get('Authorization')
        user = self.token_auth.authenticate(auth_header)

        if not user:
            return {
                'error': 'Unauthorized',
                'message': 'Valid authentication required'
            }, 401

        # Attach user to request
        request.user = user
        return None


# Example usage with Flask
def create_flask_bearer_auth():
    """Example Flask implementation"""
    from flask import request, jsonify

    # Token validator function
    def validate_token(token: str) -> Optional[Dict]:
        # In production, validate JWT or query database
        # This is a simple example
        valid_tokens = {
            'secret_token_123': {
                'user_id': 1,
                'username': 'john_doe',
                'scopes': ['read', 'write', 'admin']
            }
        }
        return valid_tokens.get(token)

    auth = BearerTokenAuth(validate_token)

    @auth.require_auth
    def protected_route(user):
        return jsonify({
            'message': f"Hello {user['username']}",
            'user_id': user['user_id']
        })

    @auth.require_scope('admin')
    def admin_route(user):
        return jsonify({'message': 'Admin panel'})

    return auth, protected_route, admin_route


# Example usage with FastAPI
def create_fastapi_bearer_auth():
    """Example FastAPI implementation"""
    from fastapi import Depends, HTTPException, status
    from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

    security = HTTPBearer()

    def validate_token(token: str) -> Optional[Dict]:
        # Token validation logic
        valid_tokens = {
            'secret_token_123': {
                'user_id': 1,
                'username': 'john_doe',
                'scopes': ['read', 'write', 'admin']
            }
        }
        return valid_tokens.get(token)

    async def get_current_user(
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ) -> Dict:
        """Dependency for authentication"""
        user = validate_token(credentials.credentials)

        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid authentication credentials",
                headers={"WWW-Authenticate": "Bearer"},
            )

        return user

    async def require_scope(required_scope: str):
        """Dependency for scope validation"""
        def scope_checker(user: Dict = Depends(get_current_user)) -> Dict:
            if required_scope not in user.get('scopes', []):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Insufficient permissions"
                )
            return user
        return scope_checker

    return get_current_user, require_scope


if __name__ == "__main__":
    # Simple example
    def simple_validator(token: str) -> Optional[Dict]:
        if token == "valid_token":
            return {'user_id': 1, 'username': 'test_user', 'scopes': ['read']}
        return None

    auth = BearerTokenAuth(simple_validator)

    # Test token extraction
    token = auth.extract_token("Bearer valid_token")
    print(f"Extracted token: {token}")

    # Test authentication
    user = auth.authenticate("Bearer valid_token")
    print(f"Authenticated user: {user}")

    # Test invalid token
    user = auth.authenticate("Bearer invalid_token")
    print(f"Invalid auth result: {user}")
